lib/core/test/unit/Cardano/Pool/RankingSpec.hs

```diff
@@ -0,0 +1,254 @@
+
+import qualified Cardano.Pool.Ranking as R
+
+spec :: Spec
+spec = do
+    describe "Stake Pool Ranking" $ do
+        describe "all else equal..." $ do
+            it "higher margin => lower desirability (or equal)"
+                $ property
+                $ allElseEqualProperty @"margin" NegativeEffect
+
+            it "higher cost => lower desirability (or equal)"
+                $ property
+                $ allElseEqualProperty @"cost" NegativeEffect
+
+            it "higher performance => higher desirability (or equal)"
+                $ property
+                $ allElseEqualProperty @"recentAvgPerformance" PositiveEffect
+
+            it "higher oversaturation => desirability NOT affected"
+                $ property prop_oversaturationDoesNotInfluenceDesirability
+
+        describe "desirability" $ do
+            it "desirability >= 0"
+                $ property prop_desirabilityNonNegative
+
+        describe "saturatedPoolRewards" $ do
+            it "a0 == 0 => saturatedPoolRewards == p*R*z0"
+                $ property prop_saturatedPoolRewardsReduces
+
+            it "saturatedPoolRewards >= 0"
+                $ property prop_saturatedPoolRewardsNonNegative
+
+        describe "mkRelativeStake" $ do
+            it "stake > R, R >= 1 => mkRelativeStake ∈ [0,1] (even with conversions)"
+                $ property prop_mkRelativeStakeBounds
+
+prop_mkRelativeStakeBounds :: EpochConstants -> Word64 -> Word64 -> Property
+prop_mkRelativeStakeBounds constants' tot stake =
+    let
+        rel = unRelativeStake
+            $ mkRelativeStake
+                (Lovelace $ fromIntegral stake)
+                constants
+    in
+        counterexample ("relative stake = " ++ show rel) $
+            tot >= 1 ==>
+                tot >= stake ==>
+                    property $ rel >= 0 && rel <= 1
+
+  where
+    constants = constants'
+        { totalRewards = Lovelace . unsafeToNonNegative $ fromIntegral tot }
+
+prop_desirabilityNonNegative :: EpochConstants -> Pool -> Property
+prop_desirabilityNonNegative constants pool = property $
+    desirability constants pool >= 0
+
+prop_saturatedPoolRewardsNonNegative :: EpochConstants -> Pool -> Property
+prop_saturatedPoolRewardsNonNegative constants pool = property $
+    saturatedPoolRewards constants pool >= 0
+
+prop_saturatedPoolRewardsReduces :: EpochConstants -> Pool -> Property
+prop_saturatedPoolRewardsReduces constants' pool =
+    let
+        constants = constants'
+            { leaderStakeInfluence = unsafeToNonNegative 0 }
+        z0 = unRelativeStake  $ saturatedPoolSize constants
+        _R = R.getNonNegative $ getLovelace $ totalRewards constants
+        p  = R.getNonNegative $ recentAvgPerformance pool
+    in
+        saturatedPoolRewards constants pool === p*_R*z0
+
+prop_oversaturationDoesNotInfluenceDesirability
+    :: Pool
+    -> EpochConstants
+    -> Property
+prop_oversaturationDoesNotInfluenceDesirability pool constants =
```

Comment:
I am not entirely convinced by what this property tests. Of course, the pool's stake doesn't influence the desirability because it's not even a parameter of the equation!

Reply:
I think it's easy to guess that `desirability` should take oversaturation into account. I wanted to make it obvious that it doesn't.

If we ever add `stake` as a field to `Pool` it will not be obvious from the type-signatures alone.